src/lib.rs

```diff
@@ -260,29 +350,82 @@ impl<B: MysqlShim<W>, R: Read, W: Write> MysqlIntermediary<B, R, W> {
                         "client sent incomplete handshake",
                     ),
-                    nom::Err::Failure((input, nom_e_kind))
-                    | nom::Err::Error((input, nom_e_kind)) => {
-                        if let nom::error::ErrorKind::Eof = nom_e_kind {
+                    nom::Err::Failure(nom_error) | nom::Err::Error(nom_error) => {
+                        if let nom::error::ErrorKind::Eof = nom_error.code {
                             io::Error::new(
                                 io::ErrorKind::UnexpectedEof,
-                                format!("client did not complete handshake; got {:?}", input),
+                                format!(
+                                    "client did not complete handshake; got {:?}",
+                                    nom_error.input
+                                ),
                             )
                         } else {
                             io::Error::new(
                                 io::ErrorKind::InvalidData,
-                                format!("bad client handshake; got {:?} ({:?})", input, nom_e_kind),
+                                format!(
+                                    "bad client handshake; got {:?} ({:?})",
+                                    nom_error.input, nom_error.code
+                                ),
                             )
                         }
                     }
                 })?
                 .1;
+
+            let mut auth_response = handshake.auth_response.clone();
+            let auth_plugin_expect = self.shim.auth_plugin_for_username(&handshake.username);
+
+            // auth switch
+            if !auth_plugin_expect.is_empty()
+                && auth_response.is_empty()
+                && handshake.auth_plugin != auth_plugin_expect.as_bytes()
+            {
+                self.writer.write_all(&[0xfe])?;
+                self.writer.write_all(auth_plugin_expect.as_bytes())?;
+                self.writer.write_all(&[0x00])?;
+                self.writer.write_all(&scramble)?;
+                self.writer.write_all(&[0x00])?;
+
+                self.writer.flush()?;
+                {
+                    let (_seq, auth_response_data) = self.reader.next()?.ok_or_else(|| {
+                        io::Error::new(
+                            io::ErrorKind::ConnectionAborted,
+                            "peer terminated connection",
+                        )
+                    })?;
+
+                    auth_response = auth_response_data.to_vec();
+                }
+            }
+
             self.writer.set_seq(seq + 1);
+
+            if !self.shim.authenticate(
+                auth_plugin_expect,
+                &handshake.username,
+                &scramble,
+                auth_response.as_slice(),
+            ) {
+                let err_msg = format!(
+                    "Authenticate failed, user: {:?}, auth_plugin: {:?}",
```

Comment:
Let's try to match the exact phrasing of the error that MySQL/MariaDB actually returns for this kind of error here.